# **Waze Project**

Welcome to the Waze Project!

Waze data analytics team is still in the early stages of their user churn project. Previously, you were asked to complete a project proposal by your supervisor, May Santner. You have received notice that your project proposal has been approved and that your team has been given access to Waze's user data. To get clear insights, the user data must be inspected and prepared for the upcoming process of exploratory data analysis (EDA).


In [1]:
# Import packages for data manipulation
### YOUR CODE HERE ###
import pandas as pd

import numpy as np

In [2]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

In [5]:
### YOUR CODE HERE ###
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [4]:
### YOUR CODE HERE ###
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [78]:
# Isolate rows with null values
missing_labels_df = df[df['label'].isnull()]  # Filter rows with missing labels

# Display summary stats of rows with null values
missing_labels_df.describe()
len(missing_labels_df)

700

In [79]:
# Isolate rows without null values
non_missing_labels_df = df[df['label'].notnull()] # Filter rows with non-missing labels

# Display summary stats of rows without null values
non_missing_labels_df.describe()
len(non_missing_labels_df)

14299

In [82]:
# Get count of null values by device
count_device = missing_labels_df['device'].value_counts()
count_device.head()

iPhone     447
Android    253
Name: device, dtype: int64

In [157]:
# Calculate % of iPhone nulls and Android nulls
percentage = missing_labels_df['device'].value_counts(normalize=True)
percentage

iPhone     0.638571
Android    0.361429
Name: device, dtype: float64

In [160]:
# Calculate % of iPhone users and Android users in full dataset
df['device'].value_counts(normalize=True)

iPhone     0.644843
Android    0.355157
Name: device, dtype: float64

The percentage of missing values by each device is consistent with their representation in the data overall.

There is nothing to suggest a non-random cause of the missing data.

In [161]:
#count of labeled records
filtered_df = df[df['label'].isin(['retained', 'churned'])].count()['label']
print ('Count of Total labeled users:', filtered_df)

#count of users who churned
churned_count = df[df['label'] == 'churned'].count()['label']
print('Count of Churned: ',churned_count)

#count of users who retained
retained_count = df[df['label'] == 'retained'].count()['label']
print('Count of Retained: ',retained_count)


print ('Percentage of Retained:', retained_count/filtered_df)
print ('Percentage of Churned:', churned_count/filtered_df)

Count of Total labeled users: 14299
Count of Churned:  2536
Count of Retained:  11763
Percentage of Retained: 0.8226449402056087
Percentage of Churned: 0.17735505979439123


In [162]:
# Calculate median values of all columns for churned and retained users

df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  km_per_drive  \
label                                                                          
churned               1607.183785            8.0           6.0     74.109416   
retained              1458.046141           17.0          14.0     75.014702   

          km_per_driving_day  drives_per_driving_day  
label                                                 
churned           697.541999                 10.0000  
retained          289.549333                  4.0625

This offers an interesting snapshot of the two groups, churned vs. retained:

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile. Continue exploring!

In [134]:
# Add a column to df called `km_per_drive`
df['km_per_drive'] = df['driven_km_drives'] /df['drives'] 
df.head()

# Group by `label`, calculate the median, and isolate for km per drive

median_values = df.groupby('label').median().iloc[:, -1]
median_values



label
churned     74.109416
retained    75.014702
Name: km_per_drive, dtype: float64

The median retained user drove about one more kilometer per drive than the median churned user. How many kilometers per driving day was this?

To calculate this statistic, repeat the steps above using `driving_days` instead of `drives`.

In [140]:
# Add a column to df called `km_per_driving_day`
df['km_per_driving_day'] = df['driven_km_drives'] /df['driving_days'] 
df.head()

# Group by `label`, calculate the median, and isolate for km per driving day
median_values = df.groupby('label').median().iloc[:, -1]
median_values

label
churned     697.541999
retained    289.549333
Name: km_per_driving_day, dtype: float64

Now, calculate the median number of drives per driving day for each group.

In [142]:
# Add a column to df called `drives_per_driving_day`
df['drives_per_driving_day'] = df['drives']/df['driving_days'] 
df.head()
# Group by `label`, calculate the median, and isolate for drives per driving day
median_values = df.groupby('label').median().iloc[:, -1]
median_values

label
churned     10.0000
retained     4.0625
Name: drives_per_driving_day, dtype: float64

The median user who churned drove 698 kilometers each day they drove last month, which is almost ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

In [165]:
# For each label, calculate the number of Android users and iPhone users

df.groupby(['label', 'device']).size()


label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [8]:
# For each label, calculate the percentage of Android users and iPhone users
df.groupby('label')['device'].value_counts(normalize=True)

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.